metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -285,42 +272,104 @@ final class FileDecoderProvider(
     } yield PathInfo(Some(targetId), classDir.resolve(relativePath))
   }
 
-  private def findTastyDirFileFromSource(
-      sourceFile: AbsolutePath,
-      position: l.Position
+  private def findPathInfoForClassesPathFile(
+      path: AbsolutePath
   ): Option[PathInfo] = {
-    for {
-      targetId <- buildTargets.sourceBuildTargets(sourceFile).headOption
-      target <- buildTargets.scalaTarget(targetId)
-      sourceRoot <- buildTargets.inverseSourceItem(sourceFile)
-      className <- classFinder.findTasty(sourceFile, position)
-    } yield {
-      val pathToTasty = className.replace('.', '/') + ".tasty"
-      val classDir = target.classDirectory.toAbsolutePath
-      val other = classDir.resolve(pathToTasty)
-      PathInfo(Some(targetId), other)
+    val pathInfos = for {
+      scalaTarget <- buildTargets.all
+      classPath = scalaTarget.classDirectory.toAbsolutePath
+      if (path.isInside(classPath))
+    } yield PathInfo(Some(scalaTarget.id), path)
+    pathInfos.toList.headOption
+  }
+
+  /**
+   * For a given scala file find all definitions (such as classes, traits, object and toplevel package definition) which
+   * may produce .class or .tasty file.
+   * If there is more than one candidate asks user, using lsp client and quickpick, which class he wants to decode.
+   * If there is only one possible candidate then just pick it.
+   *
+   * @param includeInnerClasses - if true searches for candidates which produce .class file, otherwise .tasty
+   */
+  private def findPathInfoForScalaFile(
+      path: AbsolutePath,
+      includeInnerClasses: Boolean
+  ): Future[Either[String, PathInfo]] = {
+    val availableClasses = classFinder
+      .findAllClasses(path, includeInnerClasses)
+      .filter(_.nonEmpty)
+    availableClasses match {
+      case Some(classes) =>
```

Comment:
Nit, and maybe just style preference so ignore this if it is, but it might be better to add a guard inside the match rather than `filter(_.nonEmpty)` here. The only reason I say this is that in the future someone may change this without realizing we don't guard against it in the `match`. So I'd rather leave that to the match than here. wdyt?

Reply:
It makes sense, will do